YetiForceCompany/YetiForceCRM
# WIP: Update dockerization

## Description

This PR modifies the current Dockerfile and docker-compose to:

- introduce container separation - split the previous single container to a set
- save configuration in separate volumes - avoid issues with configuration disappearing when containers are recreated
- provide easier backupability
- add caching in Dockerfile to speed up subsequent container builds

**This is a Work-In-Progress request**, so it has some things to consider before merging:

- Do we need a reverse proxy in front of YF? I included it to provide easier configurability/HTTPS support/etc., but it's possible that it may introduce some issues
- What parts of YF code should be persistent? Whenever a container is recreated, all files that were not specifically chosen will be reset - this means we could lose some config files if we don't identify them
- There are Docker-related files in /tests/setup. How do they relate to the current config?
- To gain more build speed, the Dockerfile should first copy lockfiles, then run `yarn` and `composer`, and copy all files
- Maybe we could reduce the number of volumes for the `yetiforce-crm` container

## Testing

For now, testing is manual. Try to run YF using Docker on your machine and check for possible errors.

To run dockerized YF:

1. Install [Docker Desktop](https://www.docker.com/products/docker-desktop/)
2. Run `docker compose up -d` in the repository
3. Go to `http://localhost`
4. Install YF normally, for database details enter:
  ```
  Address: db
  Port: 3306
  Username: yeti
  Password: yeti
  Database name: yetiforce
  ```
5. If you make any code changes, the following command will rebuild and restart the container: `docker compose up -d --build`

## Changes

### Final checklist
- [x] My code is written according to the guidelines found [here] (https://github.com/php-fig/fig-standards).
- [x] I have signed the Contributor Licence Agreement between me and YetiForce.
